circle-mlir/circle-mlir/lib/dialect/src/ConstFold.h

```diff
@@ -0,0 +1,263 @@
+
+  return {};
+}
+
+/// Performs const folding with broadcast behavior on the two attributes in
+/// `operands` and returns the result if possible.
+/// Depending on the given `resultType`, either `floatCalculate` or
+/// `intCalculate` is chosen to conduct the calculate.
+Attribute ConstFoldBinaryOp(Type result_type, ArrayRef<Attribute> operands,
+                            llvm::function_ref<APFloat(APFloat, APFloat)> float_calculate,
+                            llvm::function_ref<APInt(APInt, APInt)> int_calculate)
+{
+  // Note: All types are wrapped in tensor types in Circle. E.g., f32 is
+  // represented as tensor<f32>. So we are only handling tensor types here.
+  auto type = result_type.dyn_cast<ShapedType>();
+  if (!type)
+    return {};
+
+  auto elemType = type.getElementType();
+
+  if (elemType.isa<FloatType>())
+    return ConstFoldBinaryOp<FloatAttr>(result_type, operands[0], operands[1], float_calculate);
+
+  if (elemType.isSignlessInteger())
+    return ConstFoldBinaryOp<IntegerAttr>(result_type, operands[0], operands[1], int_calculate);
+
+  return {};
+}
+
+/// Performs const folding a attributes `operand` and returns the result if possible.
+/// The function currently asserts that the `result_type` to be a f32 tensor type.
+/// TODO: Extend this function to handle integral tensor for ops like "logical_not".
+Attribute ConstFoldUnaryOp(Type result_type, Attribute operand,
+                           llvm::function_ref<APFloat(APFloat)> calculate)
+{
+  assert(IsF32ShapedType(result_type));
+  auto result_shape_type = result_type.cast<ShapedType>();
+
+  if (!result_shape_type.hasStaticShape())
+    return {};
+
+  if (auto dense_elements = operand.dyn_cast_or_null<DenseElementsAttr>())
+  {
+    SmallVector<APFloat, 16> new_values;
+    const int num_elements = result_shape_type.getNumElements();
+    new_values.reserve(num_elements);
+
+    for (const APFloat &old_value : dense_elements.getValues<APFloat>())
+    {
+      new_values.push_back(calculate(old_value));
+    }
+
+    return DenseElementsAttr::get(result_shape_type, new_values);
+  }
+
+  return {};
+}
+
+} // namespace
+
+template <typename T> bool getAsConstant(mlir::Value &input, std::vector<T> &values)
+{
+  mlir::DenseElementsAttr dataAttr;
+
+  // Check if input is constant
+  if (!matchPattern(input, m_Constant(&dataAttr)))
+    return false;
+
+  if (auto constOp = dyn_cast<mlir::Circle::ConstOp>(input.getDefiningOp()))
+    dataAttr = constOp.getValueAttr().dyn_cast<mlir::DenseElementsAttr>();
+  else
+    return false;
+
+  if (dataAttr == nullptr)
+    return false;
+
+  auto valueIt = dataAttr.getValues<llvm::APInt>().begin();
+  auto valueEd = dataAttr.getValues<llvm::APInt>().end();
+  for (; valueIt != valueEd; ++valueIt)
+  {
+    T value = static_cast<T>((*valueIt).getSExtValue());
+    values.push_back(value);
+  }
+  return true;
+}
+
+bool getAsConstant(mlir::Value &input, std::vector<int64_t> &values)
```

Comment:
It would be nice to add `inline` keyword to prevent multiple definition errors.
```suggestion
inline bool getAsConstant(mlir::Value &input, std::vector<int64_t> &values)
```

Reply:
ditto